sympy/printing/pretty/stringpict.py

```diff
@@ -441,8 +441,8 @@ def __mul__(self, *others):
         if len(others) == 0:
             return self # We aren't actually multiplying... So nothing to do here.
         args = self
-        if args.binding > prettyForm.MUL:
-            arg = stringPict(*args.parens())
+        if args.binding > prettyForm.MUL and len(args[0].split()) > 1:
+            args = stringPict(*args.parens())
```

Comment:
What exactly is `args[0]` here? Is it a string?

Is this basically checking for spaces and then adding brackets if there are any?

Reply:
Yeah , Here args[0] is a string and the first one in multiplication part , so Basically I'm checking whether it is a expression or not because when there is something like `-A*1/2* we aren't adding any brackets to `A` but when we have something like 
`x + y + z*1/2` we need to add brackets to `x + y + z`.

Reply:
I think that there should be a more robust way of checking this than looking for spaces in the string. What if the string just had some spaces in it for a different reason?

I don't know the printing code so well though so I'm not sure what's best.

CC @oscargus

Reply:
I don't know the pretty printer that well, but I also think that it is a bit doubtful that it will always work. (Also, I think that `len(args[0].split()) > 1` should be replaced by `args[0].count(' ')` if this method is to be used.)

I think one should add a load of test cases if one considers this approach, including multiplication of more complicated functions that will print on multiple lines etc. What about (sin(x)/cos(x))*(1/2)? I'm also a bit surprised that the rational number is not printed as a fraction.

By the way, if I understand correctly, the existing code detects that it is multiplied by a fraction and then realizes that there is not need to add brackets since it will be written out as a fraction and you are now trying to add the brackets again. It would make more sense to approach it before it realizes that there shouldn't be any fractions, just as if it was `(x + y + z)*q`. With that said, I do not fully understand how it works.

Personally, I do not fully see the fuzz with the non-evaluated printing stuff. Much effort has been put into making the expressions print nicely, so just trying to bypass that doesn't really seem to make sense to me. But that is another question.

Reply:
Also, why is `arg` changed to `args`? It almost seems like the code wasn't working before, since `arg` was not used...

Reply:
Before , it was there as we should'nt be adding brackets to `A` whenever we multiply something like `-A*B*C` .